### PR TITLE
Added developer mode for ABP devices with flags

### PR DESCRIPTION
**Pull request by kersing**
Developer mode/relax check of Fcnt based on extensible flag.

### CODE REVIEW

**Comment by kersing** [comment]
(This replaces pull request 143 to implement #118)

File: core/components/broker/broker.go [hunk 1]
```diff
@@ -245,18 +246,27 @@ func (b component) HandleData(bctx context.Context, req *core.DataBrokerReq) (*c
 	fhdr := &uplinkPayload.MACPayload.(*lorawan.MACPayload).FHDR // No nil ref, ensured by NewLoRaWANData()
 	fcnt16 := fhdr.FCnt                                          // Keep a reference to the original counter
 
+	var fcntReset bool
 	var mEntry *devEntry
 	for _, entry := range entries {
+		fcntReset = false
```
Comment (htdvisser) [inline core/components/broker/broker.go]: `bool` is `false` by default, don't need this.

File: core/protos/handler.proto [hunk 2]
```diff
@@ -10,6 +10,7 @@ message DataUpHandlerReq {
     uint32      FCnt        = 10;
     uint32      FPort       = 11;
     uint32      MType       = 12;
+    bool	FCntUpReset = 13;
```
Comment (htdvisser) [inline core/protos/handler.proto]: Incorrect formatting

File: ttnctl/cmd/device.go [hunk 3]
```diff
@@ -75,10 +76,10 @@ registered on the Handler.`,
 
 		table := uitable.New()
 		table.MaxColWidth = 70
-		table.AddRow("DevAddr", "FCntUp", "FCntDown")
+		table.AddRow("DevAddr", "FCntUp", "FCntDown", "RelaxFcntCheck")
```
Comment (htdvisser) [inline ttnctl/cmd/device.go]: Make table header "Flags"

File: ttnctl/cmd/device.go [hunk 4]
```diff
@@ -75,10 +76,10 @@ registered on the Handler.`,
...
 		for _, device := range devices.ABP {
 			devAddr := fmt.Sprintf("%X", device.DevAddr)
-			table.AddRow(devAddr, device.FCntUp, device.FCntDown)
+			table.AddRow(devAddr, device.FCntUp, device.FCntDown, (device.Flags&core.RelaxFcntCheck) != 0)
```
Comment (htdvisser) [inline ttnctl/cmd/device.go]: Print flags as comma-separated strings

File: ttnctl/cmd/device.go [hunk 5]
```diff
@@ -196,6 +197,8 @@ var devicesInfoCmd = &cobra.Command{
 
 					fmt.Println()
 					fmt.Printf("  FCntUp:  %d\n  FCntDn:  %d\n", device.FCntUp, device.FCntDown)
+					fmt.Println()
+					fmt.Printf("  Relax Counter Checks: %t\n", device.Flags&core.RelaxFcntCheck)
```
Comment (htdvisser) [inline ttnctl/cmd/device.go]: Print flags as comma-separated strings

File: ttnctl/cmd/device.go [hunk 6]
```diff
@@ -275,7 +278,7 @@ the Handler`,
 
 // devicesRegisterPersonalizedCmd represents the `device register personalized` command
 var devicesRegisterPersonalizedCmd = &cobra.Command{
-	Use:   "personalized [DevAddr] [NwkSKey] [AppSKey]",
+	Use:   "personalized [DevAddr] [NwkSKey] [AppSKey] [RelaxFcntCheck]",
```
Comment (htdvisser) [inline ttnctl/cmd/device.go]: Remove

File: ttnctl/cmd/device.go [hunk 7]
```diff
@@ -302,6 +306,9 @@ registration on the Handler`,
 			if err != nil {
 				ctx.Fatalf("Invalid AppSKey: %s", err)
 			}
+			if (len(args) >= 4) && strings.EqualFold(args[3], "true") {
```
Comment (htdvisser) [inline ttnctl/cmd/device.go]: Just use `== "true"`

Comment (htdvisser) [inline ttnctl/cmd/device.go]: Use `cmd.Flags().GetBool("dev-mode")`

Comment (htdvisser) [inline ttnctl/cmd/device.go]: And add in `init()`:
```go
devicesRegisterPersonalizedCmd.Flags().Bool("relax-fcnt", false, "Allow frame counter to reset (insecure)")
```

**Comment by htdvisser** [comment]
Please add unit tests

File: core/components/broker/broker.go (hunk 1 above)
Comment (kersing) [inline core/components/broker/broker.go]: We're inside a loop at that point. The counter condition can set the value to true where MIC validation may continue to the next element where the variable needs to be reset.

**Comment by htdvisser** [comment]
Looks good to me. @KtorZ: Can you live with this implementation?

**Pull request merged**